scipy/sparse/_coo.py

```diff
@@ -24,82 +25,100 @@ class _coo_base(_data_matrix, _minmax_mixin):
+                self.indices = tuple(np.array(idx, copy=copy, dtype=idx_dtype)
+                                     for idx in indices)
                 self.data = getdata(obj, copy=copy, dtype=dtype)
                 self.has_canonical_format = False
         else:
             if issparse(arg1):
                 if arg1.format == self.format and copy:
-                    self.row = arg1.row.copy()
-                    self.col = arg1.col.copy()
+                    self.indices = tuple(idx.copy() for idx in arg1.indices)
                     self.data = arg1.data.copy()
-                    self._shape = check_shape(arg1.shape)
+                    self._shape = check_shape(arg1.shape, allow_1d=is_array)
+                    self.has_canonical_format = arg1.has_canonical_format
                 else:
                     coo = arg1.tocoo()
-                    self.row = coo.row
-                    self.col = coo.col
+                    self.indices = tuple(coo.indices)
                     self.data = coo.data
-                    self._shape = check_shape(coo.shape)
-                self.has_canonical_format = False
+                    self._shape = check_shape(coo.shape, allow_1d=is_array)
+                    self.has_canonical_format = False
             else:
-                #dense argument
-                M = np.atleast_2d(np.asarray(arg1))
-
-                if M.ndim != 2:
-                    raise TypeError('expected dimension <= 2 array or matrix')
-
-                self._shape = check_shape(M.shape)
+                # dense argument
+                M = np.asarray(arg1)
+                if not is_array:
+                    M = np.atleast_2d(M)
+                    if M.ndim != 2:
+                        raise TypeError('expected dimension <= 2 array or matrix')
+
+                self._shape = check_shape(M.shape, allow_1d=is_array)
                 if shape is not None:
-                    if check_shape(shape) != self._shape:
+                    if check_shape(shape, allow_1d=is_array) != self._shape:
                         message = f'inconsistent shapes: {shape} != {self._shape}'
                         raise ValueError(message)
                 index_dtype = self._get_index_dtype(maxval=max(self._shape))
-                row, col = M.nonzero()
-                self.row = row.astype(index_dtype, copy=False)
-                self.col = col.astype(index_dtype, copy=False)
-                self.data = M[self.row, self.col]
+                indices = M.nonzero()
+                self.indices = tuple(idx.astype(index_dtype, copy=False)
+                                     for idx in indices)
+                self.data = M[indices]
                 self.has_canonical_format = True
 
         if dtype is not None:
             self.data = self.data.astype(dtype, copy=False)
 
         self._check()
 
+    @property
+    def row(self):
+        return self.indices[-2] if self.ndim > 1 else np.zeros_like(self.col)
```

Comment:
Again special cased for 1/2; should we add comments where these special cases occur?